Update recommendations for ScalarDL (#213)
scalar-labs/scalar-kubernetes: docs/CreateAKSClusterForScalarDL.md

* Create the AKS cluster by using Kubernetes version 1.21 or higher.
* Configure the AKS cluster based on the version of Kubernetes and your project's requirements.

{% capture notice--info %}
**Note**

For Byzantine fault detection in ScalarDL to work properly, do not deploy your application pods on the same AKS cluster as the ScalarDL Ledger deployment.

{% endcapture %}

> <div class="notice--info">{{ notice--info | markdownify }}</div>

Reviewer: @josh-wong By the way, is there any way to preview this kind of special style somewhere or somehow (how it will be shown in the actual doc site) before merging PRs?

> <div class="notice--info">{{ notice--info | markdownify }}</div>

Reviewer 2: @jnmt Currently, previewing the docs site is a little cumbersome since the docs don't live in the same repository as where the docs site lives. For now, the method that I use to preview docs is to deploy the docs site locally with the changes and then check that the styles appear as expected.

In the near future, I'd like to have our docs based in the docs repositories so that anyone can preview docs (among other benefits).

> <div class="notice--info">{{ notice--info | markdownify }}</div>

Reviewer: > In the near future, I'd like to have our docs based in the docs repositories so that anyone can preview docs (among other benefits).

Sounds nice! It's no problem since you can preview and review changes as a final check, but nice to have it for other ones' easy review with better styles of the docs site. Thank you!

## Recommendations (optional)

The following are some recommendations for deploying ScalarDL Ledger. These recommendations are not required, so you can choose whether or not to apply these recommendations based on your needs.

### Create at least three worker nodes and three pods

To ensure that the AKS cluster has high availability, you should use at least three worker nodes and deploy at least three pods spread across the worker nodes. You can see the [sample configurations](../conf/scalardl-custom-values.yaml) of `podAntiAffinity` for making three pods spread across the worker nodes.

{% capture notice--info %}
**Note**

If you place the worker nodes in different [availability zones](https://learn.microsoft.com/en-us/azure/availability-zones/az-overview) (AZs), you can withstand an AZ failure.
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>

### Use 4vCPU / 8GB memory nodes for the worker node in the ScalarDL Ledger node pool

> From the perspective of commercial licenses, resources for one pod running ScalarDL Ledger are limited to 2vCPU / 4GB memory. In addition, some pods other than ScalarDL Ledger pods could exist on the worker nodes.

Reviewer: ```suggestion
From the perspective of commercial licenses, resources for one pod running ScalarDL Ledger are limited to 2vCPU / 4GB memory. In addition, some pods other than ScalarDL Ledger pods can exist on the worker nodes.
```

Are there any specific reasons to use the past tense? I feel 'can' is better.

Author: @jnmt  
In this sentence, what I want to tell users is `there is a possibility that some pods other than ScalarDL exist on the worker node`.

In other words, the main purpose of this sentence is NOT `we allow you to (you can) deploy some pods other than ScalarDL on the worker node`.

So, I use `could` as a `subjunctive mood` instead of `can`.

@josh-wong  
Sorry, I'm not confident about correct grammar...
What do you think?

Reviewer 2: I understand what you mean here now. The following suggestion is similar to @jnmt's suggestion, with just some rewording for active voice and a couple more questions in parentheses:

```suggestion
From the perspective of commercial licenses, resources for one pod running ScalarDL Ledger are limited to 2vCPU / 4GB memory. In addition to the ScalarDL Ledger pod, Kubernetes (or the Kubernetes service?) could allocate (all or some of?) the following pods (or components?) to each worker node:
```

If we use a sentence structure similar to the above, I think we can remove the sentence (shown below) that follows this sentence.

```markdown
In other words, the following components could run on one worker node:
```

Author: @josh-wong  

Thank you for your new suggestion!

I updated the document as follows based on your suggestion in 986ceefd8e7647ad5407edfc04ee9a900acbd4e7.

* Ledger only

  ```console
  From the perspective of commercial licenses, resources for one pod running ScalarDL Ledger are limited to 2vCPU / 4GB memory. In addition to the ScalarDL Ledger pod, Kubernetes could deploy some of the following components to each worker node:
  ```

* Auditor mode

  ```console
  From the perspective of commercial licenses, resources for each pod running ScalarDL Ledger or ScalarDL Auditor are limited to 2vCPU / 4GB memory. In addition to the ScalarDL Ledger and ScalarDL Auditor pods, Kubernetes could deploy some of the following components to each worker node:
  ```

---

The following are the answer and reason of my updates.

> Kubernetes (or the Kubernetes service?)

Strictly, pod scheduling is done by the [`kube-scheduler`](https://kubernetes.io/docs/concepts/overview/components/#kube-scheduler). But, it's a bit internal thing. So, I think `Kubernetes` is simpler and better.

> allocate

I think `allocate` is not the wrong term. But, in the Kubernetes context, I think `deploy` is better.

> (all or some of?)

It's random things. It doesn't always deploy `all` components on one worker node. So, I think `some of` is better.

> following pods (or components?)

Strictly, there are some non-pod components such as [`kubelet`](https://kubernetes.io/docs/concepts/overview/components/#kubelet) that is one of the `Kubernetes components`. So, I think `components` is better.

---

Please take a look again when you have time!

Reviewer 2: Sorry for the delay! LGTM😄

> In other words, the following components could run on one worker node:

Reviewer: Ditto, including other files.

* ScalarDL Ledger pod (2vCPU / 4GB)
* Envoy proxy
* Monitoring components (if you deploy monitoring components such `kube-prometheus-stack`)
* Kubernetes components

With this in mind, you should use a worker node that has at least 4vCPU / 8GB memory resources and use at least three worker nodes for availability, as mentioned in [Create at least three worker nodes and three pods](#create-at-least-three-worker-nodes-and-three-pods).

> However, three nodes with at least 4vCPU / 8GB memory resources per node is the minimum for a production environment. You should also consider the resources of the AKS cluster (for example, the number of worker nodes, vCPUs per node, memories per node, ScalarDL Ledger pods, and pods for your application), which depend on your system's workload. In addition, if you plan to scale the pods automatically by using some features like [Horizontal Pod Autoscaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/), you should consider the maximum number of pods on the worker node to decide on the worker node resources.

Reviewer: ```suggestion
However, three nodes with at least 4vCPU / 8GB memory resources per node is the minimum for a production environment. You should also consider the resources of the AKS cluster (for example, the number of worker nodes, vCPUs per node, memory per node, ScalarDL Ledger pods, and pods for your application), which depend on your system's workload. In addition, if you plan to scale the pods automatically by using some features like [Horizontal Pod Autoscaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/), you should consider the maximum number of pods on the worker node to decide on the worker node resources.
```

> However, three nodes with at least 4vCPU / 8GB memory resources per node is the minimum for a production environment. You should also consider the resources of the AKS cluster (for example, the number of worker nodes, vCPUs per node, memories per node, ScalarDL Ledger pods, and pods for your application), which depend on your system's workload. In addition, if you plan to scale the pods automatically by using some features like [Horizontal Pod Autoscaling (HPA)](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/), you should consider the maximum number of pods on the worker node to decide on the worker node resources.

Author: Sorry, this is my mistake. 
In case of ScalarDL, users should not deploy `user application` on the same Kubernetes cluster as ScalarDL.

Also, this document is for ScalarDL Ledger (Ledger only mode). So, the correct description is as follows.

```
(for example, the number of worker nodes, vCPUs per node, memory per node, and ScalarDL Ledger pods)
```

I fixed it in 64a19d4032cf160c9fa2269fbee20e22da8cc27a.

    * 50052/TCP (load balancing for ScalarDL Ledger)
    * 9001/TCP (accepts monitoring requests for Scalar Envoy itself)

{% capture notice--info %}
**Note**

If you change the default listening port for ScalarDL Ledger in the configuration file (`ledger.properties`), you must allow connections by using the port that you configured.
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>

{% capture notice--info %}
**Note**

> You also must allow the connections that AKS uses itself. For more details about AKS traffic requirements, refer to [Control egress traffic using Azure Firewall in Azure Kubernetes Service (AKS)](https://learn.microsoft.com/en-us/azure/aks/limit-egress-traffic).

Reviewer: ```suggestion
You must also allow the connections that AKS uses itself. For more details about AKS traffic requirements, refer to [Control egress traffic using Azure Firewall in Azure Kubernetes Service (AKS)](https://learn.microsoft.com/en-us/azure/aks/limit-egress-traffic).
```
